spec/Appendix A -- Persisted Documents.md

```diff
@@ -0,0 +1,271 @@
+the server may retrieve them given the identifier at request time. This must be
+done in a secure manner (preventing untrusted third parties from adding their
+own persisted document) such that the server will be able to retrieve the
+identified document within a _persisted document request_ and know that it is
+trusted.
+
+Note: When used solely as a bandwidth optimization, as in the technique known
+colloquially as "automatic persisted queries (APQ)," an error-based mechanism
+might be used wherein the client assumes that the document has already been
+persisted, but if the request fails due to unknown _document identifier_ the
+client issues a follow-up request containing the full GraphQL Document to be
+persisted.
+
+Note: When persisting a document for which the identifier has been derived by
+the client, it is generally good practice for the client to issue both the
+GraphQL Document and the document identifier to the server; the server could
+then regenerate the document identifier from the GraphQL Document independently,
+and check that the identifiers match before storing the Document. If the
+identifier is not derived on the client then the client must coordinate
+retrieval of a document identifier from the server to be incorporated into the
+deployed client.
+
+## Persisted Document Request
+
+A server MAY accept a _persisted document request_ via `GET` or `POST`.
+
+### Persisted Document Request Parameters
+
+:: A _persisted document request_ is an HTTP request that encodes the following
+parameters in one of the manners described in this specification:
+
+- {documentId} - (_Required_, string): The string identifier for the Document.
+- {operationName} - (_Optional_, string): The name of the Operation in the
+  identified Document to execute.
+- {variables} - (_Optional_, map): Values for any Variables defined by the
+  Operation.
+- {extensions} - (_Optional_, map): This entry is reserved for implementors to
+  extend the protocol however they see fit.
+
+### GET
+
+For a _persisted document request_ using HTTP GET, parameters SHOULD be provided
+in the query component of the request URL, encoded in the
+`application/x-www-form-urlencoded` format as specified by the
+[WhatWG URLSearchParams class](https://url.spec.whatwg.org/#interface-urlsearchparams).
+
+The {documentId} parameter must be a string _document identifier_.
+
+The {operationName} parameter, if present, must be a string.
+
+Each of the {variables} and {extensions} parameters, if used, MUST be encoded as
+a JSON string.
+
+Setting the value of the {operationName} parameter to the empty string is
+equivalent to omitting the {operationName} parameter.
+
+A client MAY provide the _persisted document request_ parameters in another way
+if the server supports that.
+
+Note: A common alternative pattern is to use a dedicated URL for each _persisted
+operation_ (e.g.
+`https://example.com/graphql/sha256:71f7dc5758652baac68e4a10c50be732b741c892ade2883a99358f52b555286b`).
+
```

Comment:
Can we add a section on i.e. returning 404 for persisted-documents we can't find and maybe even 400 if they don't leverage the right prefix?

Reply:
I think for the URL format 404 should be encouraged. I've been writing up a change to the appendix which encourages the URL format, but haven't had time to finish it yet; I've just raised a PR for my WIP so we have something to easily reference: https://github.com/graphql/graphql-over-http/pull/305

IMO for the non-URL version (traditional), 404 should not be used - it suggests that the `/graphql` endpoint is not found, which would be confusing.